File: ArrayAndStrings/insert-del-getRandom01/solution.py
import random
class RandomizedSet:

    def __init__(self):
        self.nums = []
        self.indexes = dict()

    def insert(self, val: int) -> bool:
        if val in self.indexes:
            return False
        else:
            self.nums.append(val)
            self.indexes[val] = len(self.nums)-1
            return True

    def remove(self, val: int) -> bool:
        if val in self.indexes:
            index = self.indexes[val]
            lastElement = self.nums[len(self.nums)-1]
            self.nums[index] = lastElement
            self.nums.pop()
            self.indexes[lastElement] = index
            del self.indexes[val]
            return True
        else:
            return False

    def getRandom(self) -> int:
        return random.choice(self.nums)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: ArrayAndStrings/insert-del-getRandom01/solution.js
var RandomizedSet = function () {
  this.nums = [];
  this.indexes = new Map();
};

/**
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function (val) {
  if (this.indexes.has(val)) {
    return false;
  } else {
    this.nums.push(val);
    this.indexes.set(val, this.nums.length - 1);
    return true;
  }
};

/**
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function (val) {
  if (this.indexes.has(val)) {
    const index = this.indexes.get(val);
    const lastElement = this.nums[this.nums.length - 1];
    this.nums[index] = lastElement;
    this.indexes.set(lastElement, index);
    this.nums.pop();
    this.indexes.delete(val);
    return true;
  } else {
    return false;
  }
};

/**
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function () {
  const index = Math.floor(Math.random() * this.nums.length);
  return this.nums[index];
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */
